tests/unit-tests/lib/test_str_util.cpp

```diff
@@ -39,58 +39,317 @@ namespace test_str_util {
+        EXPECT_STREQ(buf1, "[BOINC|1.2.3]");
+        EXPECT_STREQ(buf2, "");
+        EXPECT_STREQ(buf3, "[INTEL|Intel(R) HD Graphics|1|2406MB||201]");
+        strcpy(buf, "[vbox|4.5.6abc][INTEL|Intel(R) HD Graphics|1|2406MB||201][BOINC|1.2.3]");
+        parse_serialnum(buf, buf1, buf2, buf3);
+        EXPECT_STREQ(buf1, "[BOINC|1.2.3]");
+        EXPECT_STREQ(buf2, "[vbox|4.5.6abc]");
+        EXPECT_STREQ(buf3, "[INTEL|Intel(R) HD Graphics|1|2406MB||201]");
+        strcpy(buf, "[BOINC|1.2.3][INTEL|Intel(R) HD Graphics|1|2406MB||201][vbox|4.5.6abc]");
+        parse_serialnum(buf, buf1, buf2, buf3);
+        EXPECT_STREQ(buf1, "[BOINC|1.2.3]");
+        EXPECT_STREQ(buf2, "[vbox|4.5.6abc]");
+        EXPECT_STREQ(buf3, "[INTEL|Intel(R) HD Graphics|1|2406MB||201]");
+        strcpy(buf, "[BOINC|1.2.3][vbox|4.5.6abc][INTEL|Intel(R) HD Graphics|1|2406MB||201]");
+        parse_serialnum(buf, buf1, buf2, buf3);
+        EXPECT_STREQ(buf1, "[BOINC|1.2.3]");
+        EXPECT_STREQ(buf2, "[vbox|4.5.6abc]");
+        EXPECT_STREQ(buf3, "[INTEL|Intel(R) HD Graphics|1|2406MB||201]");
+        strcpy(buf, "[BOINC|7.6.22][CAL|ATI Radeon HD 5800/5900 series (Cypress/Hemlock)|2|1024MB|1.4.1848|102][vbox|5.1.26]");
+        parse_serialnum(buf, buf1, buf2, buf3);
+        EXPECT_STREQ(buf1, "[BOINC|7.6.22]");
+        EXPECT_STREQ(buf2, "[vbox|5.1.26]");
+        EXPECT_STREQ(buf3, "[CAL|ATI Radeon HD 5800/5900 series (Cypress/Hemlock)|2|1024MB|1.4.1848|102]");
+        strcpy(buf, "[BOINC|7.6.22[CAL|ATI Radeon HD 5800/5900 series (Cypress/Hemlock)|2|1024MB|1.4.1848|102][vbox|5.1.26]");
+        parse_serialnum(buf, buf1, buf2, buf3);
+        EXPECT_STREQ(buf1, "[BOINC|7.6.22[CAL|ATI Radeon HD 5800/5900 series (Cypress/Hemlock)|2|1024MB|1.4.1848|102]");
+        EXPECT_STREQ(buf2, "[vbox|5.1.26]");
+        EXPECT_STREQ(buf3, "");
+        strcpy(buf, "[BOINC|7.6.22][CAL|ATI Radeon HD 5800/5900 series [Cypress/Hemlock]|2|1024MB|1.4.1848|102][vbox|5.1.26]");
+        parse_serialnum(buf, buf1, buf2, buf3);
+        EXPECT_STREQ(buf1, "[BOINC|7.6.22]");
+        EXPECT_STREQ(buf2, "");
+        EXPECT_STREQ(buf3, "[CAL|ATI Radeon HD 5800/5900 series [Cypress/Hemlock]");
+        strcpy(buf, "[BOINC|7.6.22][CAL|ATI Radeon HD 5800/5900 series (Cypress/Hemlock)|2|1024MB|1.4.1848|102][extra|7.8.9][vbox|5.1.26]");
+        parse_serialnum(buf, buf1, buf2, buf3);
+        EXPECT_STREQ(buf1, "[BOINC|7.6.22]");
+        EXPECT_STREQ(buf2, "[vbox|5.1.26]");
+        //TODO: fix parse_serialnum so this doesn't happen:
+        EXPECT_STREQ(buf3, "[CAL|ATI Radeon HD 5800/5900 series (Cypress/Hemlock)|2|1024MB|1.4.1848|102][extra|7.8.9]");
     }
 
     TEST_F(test_str_util, is_valid_filename) {
         //char tmp = "filename.txt";
         bool ret = is_valid_filename("filename.txt");
-        ASSERT_TRUE(ret);
+        EXPECT_TRUE(ret);
         ret = is_valid_filename("../filename.txt");
-        ASSERT_FALSE(ret);
+        EXPECT_FALSE(ret);
```

Comment:
Can you add this as well.

`ret = is_valid_filename("../file\nname.txt");
EXPECT_FALSE(ret);`

This should capture all the cases for that project.

Reply:
done